cpp/src/arrow/flight/server.cc

```diff
@@ -275,78 +277,99 @@ Status FlightServerBase::GetSchema(const ServerCallContext& context,
+      }
+      // One WriteRecordBatch call might generate multiple payloads, so we
+      // need to collect them in a list.
+      RETURN_NOT_OK(writer_->WriteRecordBatch(*batch));
     }
 
-    if (stage_ == Stage::DICTIONARY) {
-      if (dictionary_index_ == static_cast<int>(dictionaries_.size())) {
-        stage_ = Stage::RECORD_BATCH;
-        return ipc::GetRecordBatchPayload(*current_batch_, ipc_options_,
-                                          &payload->ipc_message);
-      } else {
-        return GetNextDictionary(payload);
-      }
+    // There must be at least one payload generated after WriteRecordBatch or
+    // from previous calls to WriteRecordBatch.
+    if (payload_list_.empty()) {
+      return Status::UnknownError("IPC writer didn't produce any payloads");
     }
 
-    RETURN_NOT_OK(reader_->ReadNext(&current_batch_));
+    *payload = std::move(payload_list_.front());
+    payload_list_.pop_front();
+    return Status::OK();
+  }
 
-    // TODO(ARROW-10787): Delta dictionaries
-    if (!current_batch_) {
-      // Signal that iteration is over
-      payload->ipc_message.metadata = nullptr;
-      return Status::OK();
-    } else {
-      return ipc::GetRecordBatchPayload(*current_batch_, ipc_options_,
-                                        &payload->ipc_message);
+  Status Close() {
+    if (writer_) {
+      RETURN_NOT_OK(writer_->Close());
     }
+    return reader_->Close();
   }
 
-  Status Close() { return reader_->Close(); }
-
  private:
-  Status GetNextDictionary(FlightPayload* payload) {
-    const auto& it = dictionaries_[dictionary_index_++];
-    return ipc::GetDictionaryPayload(it.first, it.second, ipc_options_,
-                                     &payload->ipc_message);
-  }
+  // Simple payload writer that uses a list of generated payloads.
+  class ServerRecordBatchPayloadWriter : public ipc::internal::IpcPayloadWriter {
+   public:
+    explicit ServerRecordBatchPayloadWriter(std::list<FlightPayload>* payload_list)
+        : payload_list_(payload_list) {}
 
-  Stage stage_ = Stage::NEW;
-  std::shared_ptr<RecordBatchReader> reader_;
-  ipc::DictionaryFieldMapper mapper_;
-  ipc::IpcWriteOptions ipc_options_;
-  std::shared_ptr<RecordBatch> current_batch_;
-  std::vector<std::pair<int64_t, std::shared_ptr<Array>>> dictionaries_;
+    Status Start() override { return Status::OK(); }
 
-  // Index of next dictionary to send
-  int dictionary_index_ = 0;
+    Status WritePayload(const ipc::IpcPayload& ipc_payload) override {
+      FlightPayload payload;
+      payload.ipc_message = ipc_payload;
+
+      payload_list_->push_back(std::move(payload));
+      return Status::OK();
+    }
+
+    Status Close() override { return Status::OK(); }
+
+   private:
+    std::list<FlightPayload>* payload_list_;
+  };
+
+  std::shared_ptr<RecordBatchReader> reader_;
+  ipc::IpcWriteOptions options_;
+  std::unique_ptr<ipc::RecordBatchWriter> writer_;
+  std::list<FlightPayload> payload_list_;
```

Comment:
Ah sorry one last thing: I would prefer `std::deque` (deque) for a double-ended queue over `std::list` (linked list)